### PR TITLE
Fix of bug #1107 *Possible to Create Duplicate Resource Entries*

### LINKED ISSUE #1107: Possible to Create Duplicate Resource Entries
### Steps to reproduce the error
1. Click "Add Resource"
2. Select "Upload New"
3. Click "Choose File"
4. Select file for upload
5. Name the Resource
6. Click Save quickly a few times

### Actual behavior

The result is multiple resources. They all point to the same file on S3, but there are duplicate database entries. 

### Expected behavior

I would have expected to only be able to submit the resource 1x. After clicking submit the button should be disabled to prevent additional submissions.

### PR DESCRIPTION
**Pull request by laura-barluzzi**
### Proposed changes in this pull request

This pull request fixes #1107. Following later suggestions, this PR has exported the handler into a static JS file instead of using onclick. In order to fully fix the issue, this PR also manually: 

1. validate the form using Parsley;
2. submit the form once;
3. disable the button after the form has been submitted. 

### When should this PR be merged

No dependecies.

### Risks

The submit button may stay disabled even if processing the submission in the backend failed. To fix this, we would need to change to submitting the form via ajax. A work-around is to refresh the page.

### Follow up actions

Although the Parsley validation itself works, I noticed that the first field of the form, where you upload a file as resource, does behaves as if it's always a valid field in the parsley validation test. On the contrary, the name field return false if empty. As a follow-up action I'm happy to look at this specific field validation so that the validation test of the form `returns false` also when: 

1. No file has been uploaded;
2. The file uploaded is of an unexpected extension.

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
- [ ] **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
- [ ] **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
- [ ] **If this is a bug fix, are tests for the issue in place**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

**Comment by laura-barluzzi** [comment]
Hi @wonderchook, could you have a look at this as you wrote the original bug? Thanks.

File: cadasta/templates/resources/modal_add_lib.html [hunk 1]
```diff
@@ -33,3 +34,4 @@ <h3 class="modal-title">{% trans "Attach from library" %}</h3>
     </div>
   </div>
 </div>
+<script src="{% static 'js/form_submit_once.js' %}"></script>
```
Comment (oliverroick) [inline cadasta/templates/resources/modal_add_lib.html]: Additional scripts should be added to the `extra_scripts` block. That way we can make sure scripts are always loading in the correct order. See [here](https://github.com/Cadasta/cadasta-platform/blob/master/cadasta/templates/party/party_edit.html#L5) for an example.

File: cadasta/templates/resources/modal_upload.html [hunk 2]
```diff
@@ -35,3 +36,4 @@ <h3 class="modal-title">{% trans "Upload new resource" %}</h3>
     </div>
   </div>
 </div>
+<script src="{% static 'js/form_submit_once.js' %}"></script>
```
Comment (oliverroick) [inline cadasta/templates/resources/modal_upload.html]: Additional scripts should be added to the `extra_scripts` block. That way we can make sure scripts are always loading in the correct order. See [here](https://github.com/Cadasta/cadasta-platform/blob/master/cadasta/templates/party/party_edit.html#L5) for an example.

File: cadasta/templates/resources/modal_add_lib.html [hunk 3]
```diff
@@ -21,7 +22,7 @@ <h3 class="modal-title">{% trans "Attach from library" %}</h3>
           </div>
 
           <div class="modal-footer">
-            <button type="submit" class="btn btn-primary pull-right" name="submit">
+            <button type="submit" class="btn btn-primary pull-right" name="submit-button" onclick="return preventMultipleSubmit(this);">
```
Comment (oliverroick) [inline cadasta/templates/resources/modal_add_lib.html]: I would suggest to register the event handler into the external JS file.

File: cadasta/templates/resources/modal_upload.html [hunk 4]
```diff
@@ -23,7 +24,7 @@ <h3 class="modal-title">{% trans "Upload new resource" %}</h3>
           </div>
 
           <div class="modal-footer">
-            <button type="submit" class="btn btn-primary pull-right" name="submit">
+            <button type="submit" class="btn btn-primary pull-right" name="submit-button" onclick="return preventMultipleSubmit(this);">
```
Comment (oliverroick) [inline cadasta/templates/resources/modal_upload.html]: I would suggest to register the event handler into the external JS file.

File: cadasta/core/static/js/form_submit_once.js [hunk 5]
```diff
@@ -0,0 +1,4 @@
+function preventMultipleSubmit(button) {
```
Comment (oliverroick) [inline cadasta/core/static/js/form_submit_once.js]: I would suggest to extend the script so the registration of the event handler for the `click` event happens here as well. We're using jQuery, which makes that straightforward. You'll find a lot documentation for this online. 

What you need to do is:

1. Wait for the document to be ready.
2. Find the submit button and register a handler for the `click` event.
3. In the handler, disable the button (what you have already implemented). 

The is one more issue with that approach: We're using [Parsley](http://parsleyjs.org/) to do client-validation of the form inputs before the form is submitted to the server. When any of the inputs is invalid, users can not send the form again. (I think, you commented on this issue.) There is probably a way to find out whether the form validation failed. I would need to do some research about it, but I don't have much time at the moment because I'm working more pressing issues. Maybe you can look into that as well? This problem will need to be sorted before we can merge your PR, otherwise forms will be become unusable.

**Comment by laura-barluzzi** [comment]
@oliverroick thanks for finding the time to address all these points. I'll work on them. I just wanted to clarify why I went for the onclick option at the first place: I felt it was more immediate for a reader to understand where and when the disabling option happens. However, I'll move the event handler to the external js file :) Have a nice day!

**Comment by laura-barluzzi** [comment]
@oliverroick thank you for getting back to me. I'm happy to hear that. As a matter of fact I've been working on the form validation and Parsley right now. Simply I haven't come across a solution yet. I'm confident I'll find a way by the end of this week.

**Comment by laura-barluzzi** [comment]
Hi @bjohare, I added the feature to the forms submitting/editing locations and parties/relationships.

**Comment by bjohare** [comment]
Thanks @laura-barluzzi, but I can't seem to get this to work. I can still add multiple locations and parties by double-clicking the save button. Also, rather than merging the latest master into your branch, you should rebase your branch onto the latest master. You should hard reset your HEAD to the last good commit (20ef170acd7cc782a900d53026fa622139093afc), rebase onto master and then do a force push to this branch.

**Comment by laura-barluzzi** [comment]
Hi @bjohare, it seems there was a typo in the js file. Now it should work just fine.

**Comment by bjohare** [comment]
Hi @laura-barluzzi, this looks good. Please rebase onto the latest master and squash your commits. Thanks.

**Comment by laura-barluzzi** [comment]
Hi @bjohare  and @oliverroick, I rebased master into the branch and squashed all the commits. Please press the merge button when you get around to it :)
